Add support for logging with multiple tags

kahgoh opened the pull request:
### Description

Linked issue: #201

This change allows a single `TaggedLogger` to log to multiple tags, instead of having to create individual logger for each tag.

### Definition of Done

- [x] I read [contributing.md](./contributing.md)
- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including edge cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):

- [Tags](https://github.com/tinylog-org/tinylog/wiki/Logging#tags) will need to be updated with the addition of the new `tags` method. 

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt) (mandatory)
- [x] I agree that my GitHub user name will be published in the release notes (optional)

kahgoh committed:
Add ability to log to multiple tags

See issue #201.

pmwmedia commented:
What is the current status? When can I do the review?

kahgoh commented:
Sorry, its taking me a while. I'm still working updating the Scala API side.

pmwmedia commented:
@kahgoh Thanks for the status update :)

kahgoh commented:
I've pushed in the Scala API changes and think its ready for review now. I'll come back to add something for the documentation later when I get the chance.

pmwmedia commented:
Thank you very much for your commit :) I'm going to review it. Could you fix one compile error in the mean time?

```
/home/runner/work/tinylog/tinylog/tinylog-api/src/main/java/org/tinylog/Logger.java:[92,47] diamond operator is not supported in -source 1.6
```

For Android support, the source code still needs to be Java 6 compatible.

pmwmedia commented on tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/Logger.kt:
```diff
@@ -39,27 +39,23 @@ object Logger {
-	fun tag(tag: String?): TaggedLogger {
-		if (tag == null || tag.isEmpty()) {
-			return instance
+	fun tag(tag: String?, vararg otherTags: String?): TaggedLogger {
```
I would prefer two different methods: `tag(tag: String?)` and `tags(vararg tags: String?)`

pmwmedia commented on tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt:
```diff
@@ -21,23 +21,25 @@ import org.tinylog.provider.ProviderRegistry
+	private val debugTags = getCoveredTags(tags, Level.DEBUG)
+	private val infoTags  = getCoveredTags(tags, Level.INFO)
+	private val warnTags  = getCoveredTags(tags, Level.WARN)
+	private val errorTags = getCoveredTags(tags, Level.ERROR)
```
Actually we need booleans here. By using booleans, the JVM can detect disabled severity levels and remove these calls completely at runtime. This is the primary reason why tinylog is such fas in discarding log entries (see https://tinylog.org/v2/benchmark/#discarding). However, such optimization by the JVM won't work with sets.

Therefore, I propose to calculate the lowest severity level of all tags and store the result as boolean.

pmwmedia commented on tinylog-api-scala/src/main/scala/org/tinylog/scala/Logger.scala:
```diff
@@ -22,24 +22,27 @@ import scala.language.experimental.macros
 		*/
-	def tag(tag: String): TaggedLogger = {
-		if (tag == null || tag.isEmpty()) {
+	def tag(tag: String, otherTags: String*): TaggedLogger = {
```
I would prefer two different methods: `tag(tag: String)` and `tags(tags: String*)`

pmwmedia commented on tinylog-api/src/main/java/org/tinylog/Logger.java:
```diff
@@ -56,23 +59,44 @@ private Logger() {
 	 */
-	public static TaggedLogger tag(final String tag) {
-		if (tag == null || tag.isEmpty()) {
+	public static TaggedLogger tag(final String tag, final String...otherTags) {
```
I would prefer two different methods: `tag(final String tag)` and `tags(final String... tags)`

kahgoh replied:
I was thinking if we went with the two methods, it would be possible to call the method with no tags (e.g. `Logger.tag().info(...)`) - something that wouldn't be possible with the single method. 

If we go with the two different methods, should we return the one from the `instance` field if its called with no tags? The other possibility I could think of is return a `TaggedLogger` with no tags, but I'm not sure how meaningful that is.

pmwmedia replied:
Actually, `TaggedLogger` without tags is already in use by some users. In some use cases, an instance of tinylog's static `Logger` class is required. In this case, a `TaggedLogger` instance without any tags is the way to go.

I propose to see `Logger.tags()` as an alias for `Logger.tag(null)`. For me, it would be even legal to call `Logger.tags(null, "foo", "bar")`. In this case, the log entry will be logged as untagged log entry as well as tagged by `foo` and `bar`.

kahgoh replied:
That sounds good to me :)

pmwmedia commented on tinylog-api/src/main/java/org/tinylog/TaggedLogger.java:
```diff
@@ -35,27 +39,34 @@
+	private final Set<String> debugTags;
+	private final Set<String> infoTags;
+	private final Set<String> warnTags;
+	private final Set<String> errorTags;
```
We need booleans here (see my comment in the Kotlin implementation).

pmwmedia commented on tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt:
```diff
@@ -56,9 +58,7 @@ class TaggedLogger internal constructor(private val tag: String?) {
-		if (minimumLevelCoversTrace) {
-			provider.log(stackTraceDepth, tag, Level.TRACE, null, null, message)
-		}
+		traceTags.forEach { provider.log(stackTraceDepth, it, Level.TRACE, null, null, message) }
```
Does the the lambda call add stack trace element!? I'm not sure. Maybe we need to add `+1` to `stackTraceDepth`!?

kahgoh replied:
The call doesn't add a stack trace element, but I am thinking it might be better to refactor to use an explicit for-loop anyway.

The `forEach` call doesn't add a stack trace element because it is an [inline function](https://kotlinlang.org/docs/inline-functions.html#noinline) that inlines a for-loop (see [Kotlin documentation](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/for-each.html)). The compiler will inline a for-loop in. So, it shouldn't add a stack trace element in.

To further test this, I quickly tested this with the following code:
```kotlin
fun main(args: Array<String>) {
    doLog(1, 2, 3)
}

fun doLog(vararg values: Int) {
    values.toSet().forEach { v ->
        println(v)
        if (v == 3) {
            NullPointerException("Got 3").printStackTrace()
        }
    }
}
```

The output from this was:
```
1
2
3
java.lang.NullPointerException: Got 3
	at test.tinylog.LoopTesterKt.doLog(LoopTester.kt:13)
	at test.tinylog.LoopTesterKt.main(LoopTester.kt:6)
```

Notice the `forEach` didn't appear in the stack trace!

Anyway, this discussion was making me think about refactoring to using an explicit for-loop anyway because I think it will be easier to think about and it won't rely on `forEach` being an inline function.

pmwmedia replied:
Thank you very much for the investigation and the great explanation. Great to know that it is an inline function :) I agree with you that we should not rely on it.

pmwmedia commented on tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt:
```diff
@@ -790,11 +790,33 @@ class TaggedLogger internal constructor(private val tag: String?) {
+	 * @return {@code true} if the level is enabled for at least one of the tags in the given {@link Set}. otherwise, {@code false}
+	 */
+	private fun anyEnabled(tags: Set<String?>, level: Level): Boolean {
+		return tags.any{provider.isEnabled(stackTraceDepth, it, level)}
```
The method `anyEnabled()` adds a stack trace element to the stack trace, so `stackTraceDepth + 1` is required.

pmwmedia commented on tinylog-api-kotlin/src/test/kotlin/org/tinylog/kotlin/LoggerTest.kt:
```diff
@@ -1005,7 +996,33 @@ class LoggerTest {
+		 * set of tags with more than one tag.
+		 */
+		@Test
+		fun taggedMultiple() {
```
Good test! Could you also add an additional test for mixing tags with `null`?

pmwmedia commented:
I have just finished my review. Your code looks very well! I have found just a few minors to change.

kahgoh commented:
Thanks for the feedback and review! I have just finished and pushed the changes.

pmwmedia commented on tinylog-api-kotlin/src/main/kotlin/org/tinylog/kotlin/TaggedLogger.kt:
```diff
@@ -19,25 +19,34 @@ import org.tinylog.format.AdvancedMessageFormatter
-class TaggedLogger internal constructor(private val tag: String?) {
+class TaggedLogger internal constructor(private val tags: Set<String?>) {
+
+	constructor(tag: String?) : this(setOf(tag))
```
Do we really still need this constructor? If yes, could you declare the constructor as internal? I want to prohibit that users create their own instance.

kahgoh replied:
Nope, that constructor isn't needed anymore. I'll remove it 😉

pmwmedia commented:
Thank you very much for your changes! I could found only one last minor.

Could to push potential further changes as new separate commits? It would make the review easier for me :)

I really like your pull request! You had to change a lot of code. However, your changes are easily comprehensible.

kahgoh committed:
Remove single tag constructor from TaggedLogger.kt

This simplifies the class and prevents users from initiating their own
instances (they should be obtaining instances via the Logger).

kahgoh commented:
I've just pushed the change to remove the internal constructor

pmwmedia commented:
Thank you for your great work! I have merged your pull request with pleasure.
